src/persistent_merkle_tree/Node.zig

```diff
@@ -612,32 +616,50 @@ pub const Id = enum(u32) {
-        // This is the offset of the first shared depth between the two indices
-        var d_offset: Depth = 0; // path_len - depth;
+        // The shared depth between the previous and current index
+        // This is initialized as 0 since the first index has no previous index
+        var d_offset: Depth = 0;
 
-        const states = pool.nodes.items(.state);
-        const lefts = pool.nodes.items(.left);
-        const rights = pool.nodes.items(.right);
+        var states = pool.nodes.items(.state);
+        var lefts = pool.nodes.items(.left);
+        var rights = pool.nodes.items(.right);
 
-        // For each index specified
+        // For each index specified, maintain/update path_lefts and path_rights from root (depth 0) all the way to path_len
+        // but only allocate and update path_parents from the next shared depth to path_len
         for (0..indices.len) |i| {
             // Calculate the gindex bits for the current index
             const index = indices[i];
             const gindex: Gindex = @enumFromInt(@as(Gindex.Uint, @intCast(@intFromEnum(base_gindex) | index)));
 
-            try pool.alloc(path_parents[d_offset..path_len]);
+            // Calculate the depth offset to navigate from current index to the next
+            const next_d_offset = if (i == indices.len - 1)
+                // 0 because there is no next index, it also means node_id is now the new root
+                0
+            else
+                path_len - @as(Depth, @intCast(@bitSizeOf(usize) - @clz(index ^ indices[i + 1])));
+            if (try pool.alloc(path_parents[next_d_offset..path_len])) {
+                states = pool.nodes.items(.state);
+                lefts = pool.nodes.items(.left);
+                rights = pool.nodes.items(.right);
+            }
 
             var path = gindex.toPath();
 
             // Navigate down (to the depth offset), attaching any new updates
-            for (0..d_offset) |bit_i| {
-                if (path.left()) {
-                    path_lefts[bit_i] = path_parents[bit_i + 1];
-                } else {
-                    path_rights[bit_i] = path_parents[bit_i + 1];
+            // d_offset is the shared depth between the previous and current index so we can reuse path_lefts and path_rights up that point
+            // but update them to the path_parents to rebind starting from next_d_offset if needed
+            if (d_offset > next_d_offset) {
+                path.nextN(next_d_offset);
+                for (next_d_offset..d_offset) |bit_i| {
+                    if (path.left()) {
+                        path_lefts[bit_i] = path_parents[bit_i + 1];
+                    } else {
+                        path_rights[bit_i] = path_parents[bit_i + 1];
+                    }
+                    path.next();
                 }
-                path.next();
+            } else {
+                path.nextN(d_offset);
```

Comment:
in this case we don't need to update `path_lefts` or `path_rights` because the `rebind()` call below starts from `next_d_offset`

in the case of `d_offset == next_d_offset`, it'll be handled in the next `for` loop below